crates/astria-composer/src/executor/mod.rs

```diff
@@ -276,7 +287,101 @@ impl Executor {
+
+            info!("draining sequence actions from the executor receiver channel");
+            // drain the receiver channel
+            while let Ok(seq_action) = self.serialized_rollup_transactions.try_recv() {
+                let rollup_id = seq_action.rollup_id;
+                if let Err(e) = bundle_factory.try_push(seq_action) {
+                    warn!(
+                        rollup_id = %rollup_id,
+                        error = &e as &StdError,
+                        "failed to bundle sequence action: too large. sequence action is dropped."
+                    );
+                }
+            }
+
+            info!("draining remaining bundles from bundle factory");
+            // when shutting down, drain all the remaining bundles and submit to the sequencer
+            // to avoid any bundle loss.
+            loop {
+                let bundle = bundle_factory.pop_now();
+                if bundle.is_empty() {
+                    break;
+                }
+                match self.submit_bundle(nonce, bundle.clone()).await {
+                    Ok(new_nonce) => {
+                        debug!(
+                            bundle = %telemetry::display::json(&SizedBundleReport(&bundle)),
+                            "drained bundle successfully; new nonce: {}", new_nonce
+                        );
+                        nonce = new_nonce;
+                        total_bundles_drained_successfully += 1;
+                    }
+                    Err(e) => {
+                        error!(bundle =  %telemetry::display::json(&SizedBundleReport(&bundle)),
+                            %e, "failed submitting bundle to sequencer during shutdown; \
+                                aborting shutdown");
+                        // if we can't submit a bundle after multiple retries, we can abort
+                        // the shutdown process
+                        return Err(e);
+                    }
+                }
+            }
+
+            Ok(())
+        };
+
+        match tokio::time::timeout(Duration::from_secs(EXECUTOR_SHUTDOWN_TIME), shutdown_logic)
+            .await
+        {
+            Ok(res) => {
+                if let Err(e) = res {
+                    error!(%e, "executor shutdown tasks aborted");
+                } else {
+                    info!("executor shutdown tasks completed successfully");
+                }
+            }
+            Err(err) => {
+                error!(%err, "executor shutdown tasks failed to complete in time");
+            }
+        }
+
+        info!(
+            "{} bundles drained successfully during shutdown",
+            total_bundles_drained_successfully
+        );
+
+        // log all the bundles that have not been drained
+        loop {
+            let bundle = bundle_factory.pop_now();
+            if bundle.is_empty() {
+                break;
+            }
+
+            warn!(
```

Comment:
Please don't fire events in a hot loop. `serde::Serialize` is implemented for `[T], T: Serialize` and also for `Vec<T>: T: Serialize`, so you could just do this:
```rust
// feel free to come up with a better name ^.^
let mut undrained_bundles = Vec::new(); // potentially ::with_capacity since BundleFactory might have that info
loop {
    let bundle = bundle_factory.pop_now();
    if bundle.is_empty() {
        break;
    }
    undrained_bundles.push(bundle);
}
warn!(
    undrained_bundles = %telemetry::display::json(&undrained_bundles);
    "unable to drain all bundles within the allocated time",
);
```

Reply:
Apologies, this should have read something like this:
```rust
let report = undrained_bundles.iter().map(SizedBundleReport);
```
and then report that ^

Reply:
that works too